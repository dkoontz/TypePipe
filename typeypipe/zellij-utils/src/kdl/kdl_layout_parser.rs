// Minimal stub for KDL layout parser - functionality removed for Typey Pipe




