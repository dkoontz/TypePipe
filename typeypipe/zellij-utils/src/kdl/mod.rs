mod kdl_layout_parser;



use crate::input::config::{ConfigError};
use crate::data::{Action, PluginAliases, WebClientConfig};
use crate::input::options::{Options};








use kdl::{KdlNode};






impl Action {
    pub fn new_from_bytes(
        action_name: &str,
        bytes: Vec<u8>,
        action_node: &KdlNode,
    ) -> Result<Self, ConfigError> {
        match action_name {
            "Write" => Ok(Action::Write(None, bytes, false)),
            "PaneNameInput" => Ok(Action::PaneNameInput(bytes)),
            "TabNameInput" => Ok(Action::TabNameInput(bytes)),
            "SearchInput" => Ok(Action::SearchInput(bytes)),
            _ => Err(ConfigError::new_kdl_error(
                format!("Unsupported action with bytes: {}", action_name),
                action_node.span().offset(),
                action_node.span().len(),
            )),
        }
    }

    pub fn new_from_string(
        action_name: &str,
        string: String,
        action_node: &KdlNode,
    ) -> Result<Self, ConfigError> {
        match action_name {
            "Write" => Ok(Action::Write(Some(string), vec![], false)),
            "GoToTab" => {
                let tab_index = string.parse::<u32>().map_err(|_| {
                    ConfigError::new_kdl_error(
                        format!("Invalid tab index: {}", string),
                        action_node.span().offset(),
                        action_node.span().len(),
                    )
                })?;
                Ok(Action::GoToTab(tab_index))
            }
            _ => Err(ConfigError::new_kdl_error(
                format!("Unsupported action with string: {}", action_name),
                action_node.span().offset(),
                action_node.span().len(),
            )),
        }
    }

    pub fn to_kdl(&self) -> Option<KdlNode> {
        match self {
            Action::NoOp => {
                Some(KdlNode::new("NoOp"))
            }
            Action::Write(text, bytes, _) => {
                let mut node = KdlNode::new("Write");
                if let Some(text) = text {
                    node.push(text.clone());
                } else if !bytes.is_empty() {
                    // Convert bytes to string representation
                    node.push(format!("{:?}", bytes));
                }
                Some(node)
            }
            Action::PaneNameInput(bytes) => {
                let mut node = KdlNode::new("PaneNameInput");
                node.push(format!("{:?}", bytes));
                Some(node)
            }
            Action::TabNameInput(bytes) => {
                let mut node = KdlNode::new("TabNameInput");
                node.push(format!("{:?}", bytes));
                Some(node)
            }
            Action::SearchInput(bytes) => {
                let mut node = KdlNode::new("SearchInput");
                node.push(format!("{:?}", bytes));
                Some(node)
            }
            Action::GoToTab(tab_index) => {
                let mut node = KdlNode::new("GoToTab");
                node.push(*tab_index as i64);
                Some(node)
            }
            Action::WriteChars(chars) => {
                let mut node = KdlNode::new("WriteChars");
                node.push(chars.clone());
                Some(node)
            }
            Action::SwitchToMode(_mode) => {
                Some(KdlNode::new("SwitchToMode"))
            }
            Action::Resize(_resize, _direction) => {
                Some(KdlNode::new("Resize"))
            }
            Action::SwitchFocus => {
                Some(KdlNode::new("SwitchFocus"))
            }
            Action::ToggleTab => {
                Some(KdlNode::new("ToggleTab"))
            }
            Action::MoveFocus(_direction) => {
                Some(KdlNode::new("MoveFocus"))
            }
            Action::MoveFocusOrTab(_direction) => {
                Some(KdlNode::new("MoveFocusOrTab"))
            }
            Action::MovePane(_direction) => {
                Some(KdlNode::new("MovePane"))
            }
            Action::NewPane(_direction, _name) => {
                Some(KdlNode::new("NewPane"))
            }
            Action::NewTab(_cwd, _name) => {
                Some(KdlNode::new("NewTab"))
            }
            Action::CloseTab => {
                Some(KdlNode::new("CloseTab"))
            }
            Action::CloseFocus => {
                Some(KdlNode::new("CloseFocus"))
            }
            Action::Quit => {
                Some(KdlNode::new("Quit"))
            }
            Action::MouseEvent(_event) => {
                Some(KdlNode::new("MouseEvent"))
            }
            Action::CliPipe { pipe_id, .. } => {
                let mut node = KdlNode::new("CliPipe");
                node.push(pipe_id.clone());
                Some(node)
            }
            // Stub implementations for all other Action variants
            _ => {
                // For Phase 6 simplification, we don't need to serialize most actions to KDL
                None
            }
        }
    }
}

// Layout and Config implementations are in their respective files to avoid duplicates

// Simplified stubs for removed functionality
impl PluginAliases {
    pub fn from_kdl(_kdl_plugin_aliases: &KdlNode) -> Result<Self, ConfigError> {
        Ok(PluginAliases::default())
    }
}

impl WebClientConfig {
    pub fn from_kdl(_kdl_web_client_config: &KdlNode) -> Result<Self, ConfigError> {
        Ok(WebClientConfig::default())
    }
}

impl Options {
    pub fn from_kdl(_kdl_options: &KdlNode) -> Result<Self, ConfigError> {
        Ok(Options::default())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_config() {
        let config = Config::from_kdl("", None).unwrap();
        assert_eq!(config.options, Options::default());
    }

    #[test]
    fn test_action_serialization() {
        let action = Action::NoOp;
        let kdl_node = action.to_kdl().unwrap();
        assert_eq!(kdl_node.name().value(), "NoOp");
    }
}