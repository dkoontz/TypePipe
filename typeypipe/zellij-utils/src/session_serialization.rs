// Minimal stub for session serialization - functionality removed for Typey Pipe
use kdl::{KdlDocument, KdlEntry, KdlNode, KdlValue};
use std::collections::{BTreeMap, HashMap};
use std::path::PathBuf;

use crate::pane_size::{Constraint, PaneGeom};

#[derive(Default, Debug, Clone)]
pub struct GlobalLayoutManifest {
    pub global_cwd: Option<PathBuf>,
    pub default_shell: Option<PathBuf>,
    pub tabs: Vec<(String, TabLayoutManifest)>,
}

#[derive(Default, Debug, Clone)]
pub struct TabLayoutManifest {
    pub tiled_panes: Vec<PaneLayoutManifest>,
    pub floating_panes: Vec<PaneLayoutManifest>,
    pub is_focused: bool,
    pub hide_floating_panes: bool,
}

#[derive(Default, Debug, Clone)]
pub struct PaneLayoutManifest {
    pub geom: PaneGeom,
    pub run: Option<String>,
    pub cwd: Option<PathBuf>,
    pub is_focused: bool,
    pub pane_name: Option<String>,
    pub pane_initial_contents: Option<String>,
    pub is_borderless: bool,
    pub exclude_from_sync: bool,
}

// Stub function - returns empty layout
pub fn serialize_session_layout(
    _global_cwd: Option<PathBuf>,
    _default_shell: Option<PathBuf>,
    _tabs: Vec<(String, TabLayoutManifest)>,
    _swap_tiled_layouts: Vec<(String, String)>,
    _swap_floating_layouts: Vec<(String, String)>,
    _focused_tab_index: usize,
) -> Result<String, Box<dyn std::error::Error>> {
    Ok("// Session serialization disabled in Typey Pipe\n".to_string())
}

// Stub functions for missing session serialization functions
pub fn extract_command_and_args(_run: &str) -> (PathBuf, Vec<String>) {
    (PathBuf::from("sh"), vec![])
}

pub fn extract_edit_and_line_number(_run: &str) -> Option<(PathBuf, Option<usize>)> {
    None
}

pub fn extract_plugin_and_config(_run: &str) -> Option<(String, BTreeMap<String, String>)> {
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_serialize_empty_session() {
        let result = serialize_session_layout(None, None, vec![], vec![], vec![], 0);
        assert!(result.is_ok());
    }
}