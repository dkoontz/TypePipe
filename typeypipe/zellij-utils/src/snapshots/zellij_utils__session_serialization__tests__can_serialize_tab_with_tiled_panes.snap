---
source: zellij-utils/src/session_serialization.rs
expression: kdl.0
---
layout {
    tab name="Tab with \"tiled panes\"" {
        pane size=10
        pane cwd="/tmp/\"my/cool cwd" size=10
        pane edit="/tmp/\"my/cool cwd/my-file" size=10
        pane command="/tmp/\"my/cool cwd/command.sh" size=10 {
            start_suspended true
        }
        pane command="/tmp/\"my/cool cwd/command.sh" size=10 {
            args "--arg1" "arg\"2" "arg > \\3"
            start_suspended true
        }
        pane size=10
        pane size=10
        pane size=10 borderless=true
        pane name="my cool \\ \"pane_title\"" focus=true contents_file="initial_contents_1" size=10
    }
}

